scipy/stats/_sampling.py

```diff
@@ -735,6 +735,14 @@ def __init__(
         else:
             raise ValueError(f"Unknown distname: {distname}")
 
+    @property
+    def random_state(self):
+        return self._random_state
+
+    @random_state.setter
+    def random_state(self, random_state):
+        self._random_state = check_random_state(random_state)
```

Comment:
Is there a use case we really want to support that motivates having this as public? I would like to keep the public API as lean as possible.

Reply:
After removing `random_state` from `rvs`, I thought it is important to allow the user to change the seed easily. For example, it could be that you create the generator, then check that the rvs are indeed ok (maybe a histogram or GoF test) before you start your actual simulation with a fixed seed.
before the change, you could have done this without `self.random_state` by calling `rvs(size=..., random_state=123)` for the actual simulation

Reply:
What you describe is more like a `reset` functionality no? This is what we do in `QMCEngine.reset` for instance.

If the workflow requires to first get some samples and then run "for real", then yes this would be an issue with QMC and IMO `reset` or something else would be clearer for the user than saying update the RNG.

Reply:
Unfortunately, I think the `QMCEngine.reset()` semantics do not work well for dealing with the PRNG state ([as we've noted elsewhere](https://github.com/scipy/scipy/pull/16923#issuecomment-1234551154)). They work okay when the user passes in a static seed, but not when a user passes in a live `Generator` instance. You end up with situations where the user can easily create false correlations by unintentionally sampling the same parts of the PRNG state that they did not intend to (i.e. separate from the _intended_ resampling of the `reset()`). Or worse, not reset correctly at all.

```python
|85> rng = np.random.default_rng(0xa29cabb11cfdf44ff6cac8bec254c2a0)

|86> qrng = qmc.LatinHypercube(2, seed=rng)

|87> rng.random(4)  # Use the Generator a little after we've initialized the QMCEngine
array([0.24573226, 0.85685711, 0.02918954, 0.96401299])

|89> qrng.random(8)
array([[0.31600266, 0.17649214],
       [0.65166008, 0.89000517],
       [0.15227553, 0.38101248],
       [0.92677475, 0.10827375],
       [0.81506436, 0.79478042],
       [0.61868508, 0.73785412],
       [0.05790777, 0.29385173],
       [0.38745175, 0.57504755]])

|90> qrng.reset()
<scipy.stats._qmc.LatinHypercube at 0x7f856c4cad00>

|91> qrng.random(8)
array([[0.21928347, 0.26789286],
       [0.49635131, 0.87949838],
       [0.81600266, 0.17649214],
       [0.65166008, 0.64000517],
       [0.27727553, 0.50601248],
       [0.05177475, 0.10827375],
       [0.94006436, 0.41978042],
       [0.61868508, 0.86285412]])
```

I would probably go back to passing it in as an argument to the method. It's fine! It's a pattern that works well. It's okay to have pretty dumb classes that bundle static _data_ and behavior and not _state_ and behavior.

But if you really want this class to hold onto and manage its `Generator` state, then you do have to actually expose it like this so that the user can work with it. Making a class responsible for managing state (not just data) does have consequences for complicating its API.

Reply:
> Unfortunately, I think the QMCEngine.reset() semantics do not work well for dealing with the PRNG state (https://github.com/scipy/scipy/pull/16923#issuecomment-1234551154). They work okay when the user passes in a static seed, but not when a user passes in a live Generator instance.

I will just fix that. Somehow this went through the cracks.

Let's get some other opinions for this side of the API then. (With the assumption that the issue you pointed out would be fixed.)

Reply:
One additional point: even if this is fixed, a reset method would not be very flexible, e.g., you cannot switch from one generator to another or change the seed. so the API would not be simpler (one method vs one attribute) but less powerful.

Reply:
No, I think it's more of a fundamental problem, not an implementation issue. You can't just reset `Generator` to an earlier state safely except in very strictly controlled circumstances (that would not be appropriate to impose on a general purpose API like this one).

I'd prefer to just keep the API minimal, and pass in `random_state=` arguments.

Reply:
Ok I had another quick look. What is exactly the issue with `reset`? Because your example here does not show an issue to me. What did you expect? (Also running twice a LHS is a bad idea in general as people could be tempted to concatenate that and the result is not anymore a LHS.)

For the Halton case, I remember why there is a copy of the RNG. This is because we want the same permutation for the Van der Corput sequence (the `rng.shuffle(perm)` part.) This is a must and failing to do that result in a loss of convergence speed (cf. some articles from Art Owen and I also could verify it. I found some emails where we discussed all that when I implemented this.)

Reply:
I expected it to reset and reproduce the sequence of samples like it does elsewhere:

```python
|92> rng = np.random.default_rng(0xa29cabb11cfdf44ff6cac8bec254c2a0)

|93> qrng = qmc.LatinHypercube(2, seed=rng)

|94> qrng.random(8)
array([[0.21928347, 0.26789286],
       [0.49635131, 0.87949838],
       [0.81600266, 0.17649214],
       [0.65166008, 0.64000517],
       [0.27727553, 0.50601248],
       [0.05177475, 0.10827375],
       [0.94006436, 0.41978042],
       [0.61868508, 0.86285412]])

|95> qrng.reset()
<scipy.stats._qmc.LatinHypercube at 0x7f856c1a7b50>

|96> qrng.random(8)
array([[0.21928347, 0.26789286],
       [0.49635131, 0.87949838],
       [0.81600266, 0.17649214],
       [0.65166008, 0.64000517],
       [0.27727553, 0.50601248],
       [0.05177475, 0.10827375],
       [0.94006436, 0.41978042],
       [0.61868508, 0.86285412]])
```

I will leave the use case of having a `reset()` method at to you; I assume there is a use case for `reset()` since you implemented it and are advocating for it again here. I'm just pointing out that it fails to do what it claims in this case.

We can talk about the details of the `Halton` case elsewhere (I don't think it's true that the permutations are the same because they are of sequences of different sizes), but even if you do need N identical parallel streams to pass to N workers, you can't just copy the original PRNG state N times to get that because the original stream won't advance. That means that you are going to be resampling the same parts of the PRNG stream even for places where you didn't want that (i.e. between a worker and the original caller). And that's the general problem with any kind of resetting semantics on top of a PRNG. You make it really easy for the user to shoot themselves in the foot. It's actually safer for them to handle the `Generator` themselves rather than try to hide it from them.

If you want a simple, minimal API, let users pass in the `random_state=` as an argument when they call `rvs()`. Any kind of PRNG state management is a complication that we don't need here.

Reply:
You will have to be more specific. So far you showed examples of things I am expecting to see. In the last example the sample is the same after the reset. And before that yes there is a difference because you advanced the sample. And one LHS cannot be compared to another smaller or bigger one and won't have the same elements in them. So I don't see what you want to point out with these.

For Halton, the sequence are of the same size.

But yes, let's continue this part of the conversation on an issue if you think there is an issue.

Reply:
Yes, we can talk about the details of `QMCEngine.reset()` in other issues. I do strongly contend that they are blockers for adding similar `reset()` methods here. In any case, it doesn't seem like anyone else has any desire for that style of API here (whether or not it can be implemented correctly and safely). No one else wants this class to take on the responsibility of managing that state.

If you object to exposing the stored `.random_state` attribute here in favor of a truly minimal API, that's fine. I recommend keeping it in the method arguments like most other things that randomly sample.

Reply:
> it doesn't seem like anyone else has any desire for that style of API here (whether or not it can be implemented correctly and safely). No one else wants this class to take on the responsibility of managing that state.

Let's not jump to conclusions that quickly. This thread is barely a day old. In general this PR will need to see more eyes.

I am still -1 for having `random_state` and `reset` was just a proposal (and there might be other options) if you really want to give some sort of control to users. Which I am still not convinced is needed.

Reply:
I will write an email to the mailing list this weekend to ask for further opinions.
I see mainly two options: 

- have a public attribute 
- add random_state as a keyword to rvs, 

I just think it is important to have an API that allows users to change random_state. while not every user will need this, I don't think it complicates the API a lot, one can simply ignore this optional parameter.

Reply:
and good to see that the discussion helped to find a bug in SciPy :)